chase: return absolute path when dir_fd points to the root directory

yuwata opened the pull request.

DaanDeMeyer commented on src/basic/chase.c:
```diff
@@ -550,19 +550,20 @@ int chase(const char *path, const char *root, ChaseFlags flags, char **ret_path,
+                        char *q;
 
-                        q = path_join(root, p);
+                        /* When "root" points to the root directory, the result of chaseat() is always
+                         * absolute, hence it is not necessary to prefix with the root. When "root" points to
+                         * a non-root directory, the result path is always normalized and relative, hence
+                         * we can simply call path_join() and not necessary to call path_simplify().
+                         * Note that the result of chaseat() may start with "." (more specifically, it may be
+                         * "." or "./"), and we need to drop "." in that case. */
```
Shouldn't we change this so that we don't prefix with root if `chaseat()` returns an absolute path? So that if `CHASE_AT_RESOLVE_IN_ROOT` is not set and we follow an absolute symlink, we don't try to prefix that symlink with the root?

yuwata replied:
Sorry, I cannot follow.

> Shouldn't we change this so that we don't prefix with root if `chaseat()` returns an absolute path?

With this change, we do not prefix with root when we get absolute path from `chaseat()`.

> So that if `CHASE_AT_RESOLVE_IN_ROOT` is not set and we follow an absolute symlink, we don't try to prefix that symlink with the root?

`chase()` unconditionally (superficially not, but effectively so) set `CHASE_AT_RESOLVE_IN_ROOT`. Note, `CHASE_AT_RESOLVE_IN_ROOT` is meaning less if `root` is `"/"`.
Hence, `chaseat()` returns relative path when `root` is not `"/"`.

yuwata replied:
Added assertions about the returned path from `chaseat()`. PTAL.

DaanDeMeyer replied:
Let's say we dir_fd points to /abc/def and we resolve dir_fd, "qed" where qed is an absolute symlink to /usr. Now chaseat() will return an absolute path (see your earlier commit) and `root` will not be `NULL` or "/" in chase(), so an assertion will be triggered.

yuwata replied:
In that case, `chaseat()` returns an absolute path only when `CHASE_AT_RESOLVE_IN_ROOT` is _not_ set.

If `CHASE_AT_RESOLVE_IN_ROOT` is set, the result will be `"usr"`, instead of `"/usr"`.

mrc0mmand commented:
```
[   39.035608] Freeing unused decrypted memory: 2036K
[   39.119629] Freeing unused kernel image (initmem) memory: 2532K
[   39.122297] Write protecting the kernel read-only data: 24576k
[   39.129307] Freeing unused kernel image (text/rodata gap) memory: 2012K
[   39.133618] Freeing unused kernel image (rodata/data gap) memory: 1884K
[   39.238689] systemd[1]: Assertion 'path_is_absolute(p)' failed at src/basic/chase.c:563, function chase(). Aborting.
[   39.242781] traps: systemd[1] general protection fault ip:7f6979e83f81 sp:7fffcb51e400 error:0 in libc-2.28.so[7f6979e62000+1bb000]
[   39.247911] Kernel panic - not syncing: Attempted to kill init! exitcode=0x0000000b
[   39.247911] 
[   39.248070] CPU: 1 PID: 1 Comm: systemd Not tainted 4.18.0-408.el8.x86_64 #1
[   39.248070] Hardware name: Red Hat KVM, BIOS 1.16.0-3.module_el8.7.0+1218+f626c2ff 04/01/2014
[   39.248070] Call Trace:
[   39.248070]  dump_stack+0x41/0x60
[   39.248070]  panic+0xe7/0x2ac
[   39.248070]  do_exit.cold.25+0x43/0x96
[   39.248070]  do_group_exit+0x3a/0xa0
[   39.248070]  get_signal+0x158/0x860
[   39.248070]  ? general_protection+0x8/0x30
[   39.248070]  do_signal+0x36/0x690
[   39.248070]  ? kmem_cache_alloc+0x13f/0x280
[   39.248070]  ? general_protection+0x8/0x30
[   39.248070]  exit_to_usermode_loop+0x89/0xf0
[   39.248070]  prepare_exit_to_usermode+0x9b/0xa0
[   39.248070]  retint_user+0x8/0x8
[   39.248070] RIP: 0033:0x7f6979e83f81
[   39.248070] Code: 03 75 14 c7 05 7c ed 39 00 04 00 00 00 bf 06 00 00 00 e8 52 ca 02 00 83 3d 6b ed 39 00 04 75 0b c7 05 5f ed 39 00 05 00 00 00 <f4> 83 3d 57 ed 39 00 05 75 14 c7 05 4b ed 39 00 06 00 00 00 bf 7f
[   39.248070] RSP: 002b:00007fffcb51e400 EFLAGS: 00000246
[   39.248070] RAX: 0000000000000000 RBX: 0000000000000004 RCX: 0000000000000000
[   39.248070] RDX: 0000000000000000 RSI: 00007fffcb51e2e0 RDI: 0000000000000002
[   39.248070] RBP: 00007fffcb51e530 R08: 0000000000000000 R09: 00007fffcb51e2e0
[   39.248070] R10: 0000000000000008 R11: 0000000000000246 R12: 0000000000000001
[   39.248070] R13: 00007f697a504d9b R14: 0000000000000001 R15: 0000000000000010
[   39.248070] Kernel Offset: 0x7800000 from 0xffffffff81000000 (relocation range: 0xffffffff80000000-0xffffffffbfffffff)
```

yuwata commented:
Added three more commits.
- (hopefully) fixed https://github.com/systemd/systemd/pull/27299#issuecomment-1511403680
- fixed https://github.com/systemd/systemd/pull/27299#issuecomment-1511403680

The change is slightly bigger. I do not set the green label. Please take a look!

mrc0mmand commented:
> Added three more commits.
> 
>     * (hopefully) fixed [chase: return absolute path when dir_fd points to the root directory #27299 (comment)](https://github.com/systemd/systemd/pull/27299#issuecomment-1511403680)
> 
>     * fixed [chase: return absolute path when dir_fd points to the root directory #27299 (comment)](https://github.com/systemd/systemd/pull/27299#issuecomment-1511403680)
> 
> 
> The change is slightly bigger. I do not set the green label. Please take a look!

Unfortunately, a new issue popped up:

```
[   15.958214] Freeing unused kernel image (initmem) memory: 2532K
[   15.964290] Write protecting the kernel read-only data: 24576k
[   15.971207] Freeing unused kernel image (text/rodata gap) memory: 2012K
[   15.975359] Freeing unused kernel image (rodata/data gap) memory: 1884K
[   16.086935] systemd[1]: Failed to determine whether /proc is a mount point: Function not implemented
[   16.088846] systemd[1]: Failed to determine whether /sys is a mount point: Function not implemented
[   16.090297] systemd[1]: Failed to determine whether /dev is a mount point: Function not implemented
[[0;1;31m!!!!!![0m] Failed to mount early API filesystems.
[   16.104456] systemd[1]: Freezing execution.
```

yuwata commented:
Yay, CentOS CIs are now all green again!

DaanDeMeyer commented on src/basic/chase.h:
```diff
@@ -42,6 +42,8 @@ bool unsafe_transition(const struct stat *a, const struct stat *b);
 
 int chase(const char *path_with_prefix, const char *root, ChaseFlags chase_flags, char **ret_path, int *ret_fd);
 
+int prefix_path_from_chaseat(const char *path, const char *root, char **ret);
```
`chaseat_prefix_root()` maybe?

DaanDeMeyer commented on src/basic/chase.c:
```diff
@@ -583,6 +583,37 @@ int chase(const char *path, const char *root, ChaseFlags flags, char **ret_path,
         return r;
 }
 
+int prefix_path_from_chaseat(const char *path, const char *root, char **ret) {
```
Shouldn't we use this is chase() as well?

yuwata replied:
I do not think so, calling `path_make_absolute_cwd()` and `path_simplify()` is not necessary in `chase()`, as it is already done at the beginning of `chase()` in slightly different and complex conditions.

keszybz commented:
OK, it also makes `bootctl --print-{esp,boot}-path` work again. Can we please have a test for those two and `--print-root-device` in one of the integration tests?

yuwata commented on src/basic/fd-util.c:
```diff
@@ -911,10 +911,21 @@ int dir_fd_is_root(int dir_fd) {
                 int mntid;
 
                 r = path_get_mnt_id_at(dir_fd, "", &mntid);
+                if (r == -ENOSYS) /* The kernel is too old and /proc is not mounted yet? Let's assume that
+                                   * the file system does not have the above trick. */
+                        return true;
```
Ouch, indeed.

yuwata replied:
This happens only when an old kernel is used without /proc mounted, i.e. an early boot stage that we can assume that there is no tricky mount point. Of course, still user can setup a mount namespace without /proc mounted and construct a tricky bind mount, but I do not think it is worth to support such case.

yuwata commented:
@keszybz test is added. PTAL.

SuperSandro2000 commented on src/basic/chase.c:
```diff
@@ -532,19 +551,27 @@ int chase(const char *path, const char *original_root, ChaseFlags flags, char **
+                         * "." or "./"), and we need to drop "." in that case. */
+
+                        if (empty_or_root(root))
+                                assert(path_is_absolute(p));
```
I've just run into the following assert when someone I helped mounted two zfs datasets at /. Not sure if this falls under bug or wrong usage.

mrc0mmand replied:
Could you, please, file a new issue with some more details? So we can look into what's going on.

SuperSandro2000 replied:
#29559 I hope that's not to vague 😅
